drivers/flash/soc_flash_nrf5.c

```diff
@@ -165,8 +201,336 @@ static int nrf5_flash_init(struct device *dev)
 {
 	dev->driver_api = &flash_nrf5_api;
 
+	k_sem_init(&sem_lock, 1, 1);
+
+#if defined(CONFIG_SOC_FLASH_NRF5_RADIO_SYNC)
+	k_sem_init(&sem_sync, 0, 1);
+#endif /* CONFIG_SOC_FLASH_NRF5_RADIO_SYNC */
+
 	return 0;
 }
 
 DEVICE_INIT(nrf5_flash, CONFIG_SOC_FLASH_NRF5_DEV_NAME, nrf5_flash_init,
 	     NULL, NULL, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
+
+#if defined(CONFIG_SOC_FLASH_NRF5_RADIO_SYNC)
+
+static void time_slot_callback_work(u32_t ticks_at_expire, u32_t remainder,
+		u16_t lazy,
+		void *context)
+{
+	struct flash_op_desc *op_desc;
+	u8_t instance_index;
+	u8_t ticker_id;
+	int result;
+	u32_t err;
+
+	err = ll_radio_state_is_idle();
+	__ASSERT(!err, "Radio is on during flash operation.\n");
+
+	op_desc = context;
+
+	if (op_desc->handler(op_desc->context) == FLASH_OP_DONE) {
+		ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+		/* Stop the time slot ticker */
+		result = ticker_stop(instance_index,
+			  0,
+			  ticker_id,
+			  NULL,
+			  NULL);
+
+		if ((result != TICKER_STATUS_SUCCESS) &&
+		    (result != TICKER_STATUS_BUSY)) {
+			__ASSERT(0, "Failed to stop ticker.\n");
+		}
+
+		((struct flash_op_desc *)context)->result = 0;
+
+		/* notify thread that data is available */
+		k_sem_give(&sem_sync);
+	}
+}
+
+static void time_slot_callback_helper(u32_t ticks_at_expire, u32_t remainder,
+		u16_t lazy, void *context)
+{
+	u8_t instance_index;
+	u8_t ticker_id;
+	int err;
+
+	ll_radio_state_abort();
+
+	ll_timeslice_ticker_id_get(&instance_index, &ticker_id);
+
+	/* start a secondary one-shot ticker after ~ 500 us, */
+	/* this will let any radio role to gracefully release the Radio h/w */
+
+	err = ticker_start(instance_index, /* Radio instance ticker */
+		0, /* user_id */
+		0, /* ticker_id */
+		ticks_at_expire, /* current tick */
+		TICKER_US_TO_TICKS(FLASH_RADIO_ABORT_DELAY_US), /* 1st int. */
+		0, /* periodic (on-shot) */
+		0, /* per. remaind. (on-shot) */
+		0, /* lazy, voluntary skips */
+		0,
+		time_slot_callback_work, /* handler for executing */
+					 /* the flash operation */
+		context, /* the context for the flash operation */
+		NULL, /* no op callback */
+		NULL);
+
+	if ((err != TICKER_STATUS_SUCCESS) && (err != TICKER_STATUS_BUSY)) {
```

Comment:
Minor thing, but the parentheses around `(a != b)` is unnecessary, i.e. it can be `if (a != b && c != d)`